packages/sync-actions/src/utils/create-build-actions.js

```diff
@@ -1,7 +1,57 @@
+import isEqual from 'lodash.isequal'
+
 function applyOnBeforeDiff(before, now, fn) {
   return fn && typeof fn === 'function' ? fn(before, now) : [before, now]
 }
 
+function setPriceId(newPrice, oldVariantArray) {
+  let newPriceId = ''
+  const comparisonNewPrice = {
+    value: { currencyCode: newPrice.value.currencyCode },
+    channel: newPrice.channel,
+    country: newPrice.country,
+    customerGroup: newPrice.customerGroup,
```

Comment:
`validFrom` and `validUntil` is also part of price scope. There are some use-cases that do not seem to be covered here. We are working on same feature for Java currently and we use this [spec](https://wiki.commercetools.com/display/PS/Product+Variant+prices+sync) as base for implementation. Maybe we could sync together? FYI: @heshamMassoud @andrii-kovalenko-ct

Reply:
[Add Price](https://docs.commercetools.com/http-api-projects-products.html#add-price) says:

Adding a price is rejected:
  - if the product already contains a price with the same price scope (same currency, country, customer group and channel).
  - if **two prices have validity periods that overlap within the same price scope**. A price without validity period does not conflict with a price defined for a time period.

So, matching prices by only currency-country-channel-group is not enough, because each unique scope could have multiple non-overlapping validity periods.

Reply:
Aha! Thanks a lot @butenkor, I didn't know this. Will add those to the scope.

Reply:
So, just to be clear, it's safe to generate a `changePrice` action if a price has overlapping validity in the same scope (as a new price cannot be added), even if this validity cannot be updated (No `changeValidFrom`/`changeValidUntil`/`setValidFrom`/`setValidUntil` update actions exist)

Reply:
@butenkor Please have another look now. I added `validFrom` and `validUntil` to the scope and some more tests to verify that it works.

Reply:
having 2 separate prices for _Jan_ and _Feb_ in **`UK-GBP`** :
```json
[
  {
    "value": {
      "currencyCode": "GBP",
      "centAmount": 123
    },
    "id": "416d1f01-e17e-41e7-abbd-dc9c6167b273",
    "country": "UK",
    "validFrom": "2018-01-01T00:00:00.000Z",
    "validUntil": "2018-01-31T00:00:00.000Z"
  },
  {
    "value": {
      "currencyCode": "GBP",
      "centAmount": 123,
    },
    "id": "f11a1159-9827-4793-b6ec-31791d2b9511",
    "country": "UK",
    "validFrom": "2018-02-01T00:00:00.000Z",
    "validUntil": "2018-02-28T00:00:00.000Z"
  }
]
```

We **can**: decrease _Jan_ price validity, for instance make it valid only _01-15 Jan_:
```json
{
   "actions": [
    {
      "action": "changePrice",
      "priceId": "416d1f01-e17e-41e7-abbd-dc9c6167b273",
      "price": {
        "value": {
          "currencyCode": "GBP",
          "centAmount": 123
        },
        "country": "UK",
        "validFrom": "2018-01-01",
        "validUntil": "2018-01-15"
      }
    }
  ]
}
```

We **can NOT**: increase _Jan_ price validity to _Feb_, like this:
```json
{
  "actions": [
    {
      "action": "changePrice",
      "priceId": "416d1f01-e17e-41e7-abbd-dc9c6167b273",
      "price": {
        "value": {
          "currencyCode": "GBP",
          "centAmount": 123
        },
        "country": "UK",
        "validFrom": "2018-01-01",
        "validUntil": "2018-02-15"
      }
    }
  ]
}
```
this causes error with message:

> The value '{\"overlappingPrices\":[{\"value\":{\"type\":\"centPrecision\",\"currencyCode\":\"GBP\",\"centAmount\":123,\"fractionDigits\":2},\"id\":\"416d1f01-e17e-41e7-abbd-dc9c6167b273\",\"country\":\"UK\",\"validFrom\":\"2018-01-01T00:00:00.000Z\",\"validUntil\":\"2018-02-15T00:00:00.000Z\"},{\"value\":{\"type\":\"centPrecision\",\"currencyCode\":\"GBP\",\"centAmount\":123,\"fractionDigits\":2},\"id\":\"f11a1159-9827-4793-b6ec-31791d2b9511\",\"country\":\"UK\",\"validFrom\":\"2018-02-01T00:00:00.000Z\",\"validUntil\":\"2018-02-28T00:00:00.000Z\"}]}' is not valid for field 'price'. Two prices have overlapping validity periods."

So, **it's not safe to generate `changePrice` action if the new `PriceDraft` contains overlapping validity date in the same scope**.

Reply:
This module should identify the prices by `id` or by `scope` and generate actions for it. If you want to update an existing price (by `id`) to another price's scope, the action will be generated but will fail on the API. I think failing on the API is the better way as choosing not to generate the action at all requires an overly complex logic; also including the case where the id is passed in

Reply:
it is a responsibility of prices source/supplier: they should supply prices with valid scopes and validity period.
But in the same time we have explicitly specify what happens to existing prices on the target project. In our [Prices Sync](https://wiki.commercetools.com/display/PS/Product+Variant+prices+sync) documentation we explicitly specify that all the prices which don't exist in the _source_ set will be removed from _target_! This means during the sync you can't just add new prices or change them: you always obligated to supply **full prices set**.

But if you implement a service which expects to just add/update prices from some prices diff - you should somehow manage the case, when new prices in the source are valid and are not overlapping, but some existent prices on target overlap with incoming prices. 
The easiest way, of course, just to create update action and fail on its execution, but at least you have to specify this case as a known service behavior.

Reply:
Currently, if a price is not matched, a `removePrice` action is generated. So the price with overlapping scope, if not in the product to be imported will be removed. This error should only be generated if there are two prices with overlapping scope **to be imported** from the same file.
Sorry I missed that implementation detail in my earlier post @andrii-kovalenko-ct

Reply:
ok, this sounds reasonable: 
***the error will be thrown only in case the new source set has overlapping prices*** 
e.g. if the source supplier is corrupted.